### PR TITLE
[SPARK-9452] [SQL] Support records larger than page size in UnsafeExternalSorter

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch extends UnsafeExternalSorter to support records larger than the page size. The basic strategy is the same as in #7762: store large records in their own overflow pages.

### CODE REVIEW

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java [hunk 1]
```diff
@@ -345,26 +338,62 @@ public void insertRecord(
...
-      currentPagePosition,
+      dataPagePosition,
       lengthInBytes);
-    currentPagePosition += lengthInBytes;
-    freeSpaceInCurrentPage -= totalSpaceRequired;
     inMemSorter.insertRecord(recordAddress, prefix);
+
+    // --- Update bookeeping data structures ------------------------------------------------------
+    if (!useOverflowPage) {
+      freeSpaceInCurrentPage -= totalSpaceRequired;
```
Comment (rxin) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: should we move this into the else branch earlier? seems more clear there

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java [hunk 2]
```diff
@@ -378,32 +407,72 @@ public void insertRecord(
...
     PlatformDependent.copyMemory(
-      valueBaseObj, valueOffset, dataPageBaseObject, currentPagePosition, valueLen);
-    currentPagePosition += valueLen;
+      valueBaseObj, valueOffset, dataPageBaseObject, dataPagePosition, valueLen);
 
-    freeSpaceInCurrentPage -= totalSpaceRequired;
     inMemSorter.insertRecord(recordAddress, prefix);
+
+    // --- Update bookeeping data structures ------------------------------------------------------
+    if (!useOverflowPage) {
```
Comment (rxin) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: same here -- maybe makes more sense to put them in the else branch?

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java [hunk 3]
```diff
@@ -378,32 +407,72 @@ public void insertRecord(
...
+      freeSpaceInCurrentPage -= totalSpaceRequired;
+      currentPagePosition += totalSpaceRequired;
+    }
   }
 
+  /**
+   * Returns a sorted iterator. It is the caller's responsibility to call `freeMemory()` after
```
Comment (rxin) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: is it just freeMemory, or cleanUpResources?

**Comment by rxin** [comment]
High level approach looks good to me.

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java (hunk 3 above)
Comment (JoshRosen) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: It looks like `UnsafeExternalSorter` has separate `freeMemory()` and `deleteSpillFiles()` methods. I can add a `cleanUpResources` method to perform both of these cleanup tasks.

**Comment by JoshRosen** [comment]
Pushed a partial revision now; still going to add another test plus will consolidate cleanup methods, but I want this set of changes to run with Jenkins while I'm on the train.

**Comment by rxin** [comment]
LGTM

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Thanks - I've merged this.

**Pull request closed without merging**